nixd/lib/AST/AttrLocator.cpp

```diff
@@ -0,0 +1,36 @@
+#include "nixd/AST/AttrLocator.h"
+#include "nixd/AST/ParseAST.h"
+#include "nixd/Expr/Expr.h"
+
+namespace nixd {
+
+AttrLocator::AttrLocator(const ParseAST &AST) {
+  // Traverse the AST, record all attr locations
+  // We can assume that Nix attrs are not interleaving
+  // i.e. they are concrete tokens.
+  struct VTy : RecursiveASTVisitor<VTy> {
+    AttrLocator &This;
+    const ParseAST &AST;
+    bool visitExprAttrs(const nix::ExprAttrs *EA) {
+      for (const auto &[Symbol, Attr] : EA->attrs) {
+        lspserver::Range Range = AST.nPair(Attr.pos);
+        // Note: "a.b.c" will be assigned more than once
+        // Our visitor model ensures the order that descendant nodes will be
+        // visited at last, for nesting attrs.
+        This.Attrs[Range] = Attr.e;
+      }
+      return true;
+    }
+  } V{.This = *this, .AST = AST};
+  V.traverseExpr(AST.root());
+}
+
+const nix::Expr *AttrLocator::locate(lspserver::Position Pos) const {
+  auto It = Attrs.lower_bound({Pos, Pos});
```

Comment:
```suggestion
  auto It = Attrs.upper_bound({Pos, {INT_MAX, INT_MAX}});
```

special case for `Pos == *It`

Reply:
maybe we should check if the range actually covers the `Pos`, otherwise return nullptr.